generator/src/main/java/org/apache/kafka/message/Versions.java

```diff
@@ -118,6 +124,53 @@ public Versions intersect(Versions other) {
         return new Versions(newLowest, newHighest);
     }
 
+    /**
+     * Return a new version range that trims some versions from this range, if possible.
+     * We can't trim any versions if the resulting range would be disjoint.
+     *
+     * Some examples:
+     * 1-4.trim(1-2) = 3-4
+     * 3+.trim(4+) = 3
+     * 4+.trim(3+) = none
+     * 1-5.trim(2-4) = null
+     *
+     * @param other                 The other version range.
+     * @return                      A new version range.
+     */
+    public Versions subtract(Versions other) {
+        if (other.lowest() <= lowest) {
+            if (other.highest >= highest) {
+                // Case 1: other is a superset of this.  Trim everything.
+                return Versions.NONE;
+            } else if (other.highest < lowest) {
+                // Case 2: other is a disjoint version range that is lower than this.  Trim nothing.
+                return this;
+            } else {
+                // Case 3: trim some values from the beginning of this range.
+                //
+                // Note: it is safe to assume that other.highest() + 1 will not overflow.
+                // The reason is because if other.highest() were Short.MAX_VALUE,
+                // other.highest() < highest could not be true.
+                return new Versions((short) (other.highest() + 1), highest);
+            }
+        } else if (other.highest >= highest) {
+            int newHighest = other.lowest - 1;
+            if (newHighest < 0) {
+                // Case 4: other was NONE.  Trim nothing.
+                return this;
+            } else if (newHighest < highest) {
+                // Case 5: trim some values from the end of this range.
+                return new Versions(lowest, (short) newHighest);
+            } else {
+                // Case 6: other is a disjoint range that is higher than this.  Trim nothing.
+                return this;
+            }
+        } else {
+            // Case 7: the difference between this and other would be two ranges, not one.
+            return null;
```

Comment:
It's not necessarily illegal -- we sometimes want to handle this case.  I could create a special exception for this, but returning null seemed simpler.

Reply:
I think I'd prefer something other than `null` here. It looks like Versions are copied around to a few places where we would need to worry about NPEs. `Optional` feels overkill, so maybe just another special Versions instance like `INVALID` or something?

Reply:
There isn't a `Versions#INVALID` and I don't think we should create one.  All Versions objects currently represent a version range conceptually, and I think that's a good property to keep.  There is a `Versions#NONE`, but that isn't what we want here (it is the empty range).

I think the choices are basically returning `null`, returning `None`, or throwing an exception.  

If we do choose `null`, we don't actually have to check for it since Java already does everywhere.  I can see the argument for returning `Some / None`, though, since it makes it clearer that this function may return nothing.  In practice, most of the time when we call `subtract`, we know that it will return a valid result, so I don't bother checking.  Java will throw an NPE if a precondition was violated somewhere.

I did consider implementing collections of ranges so that this function could be defined over all inputs.  But in practice, this massively increased the complexity of the Versions code for no benefit in practice.

Reply:
Understood. Thanks for the clarification